contracts/zero-ex/contracts/src/features/libs/LibSignedCallData.sol

```diff
@@ -0,0 +1,67 @@
+/*
+
+  Copyright 2020 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.6.5;
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-utils/contracts/src/v06/LibBytesV06.sol";
+
+
+/// @dev Library for working with signed calldata.
+library LibSignedCallData {
+    using LibBytesV06 for bytes;
+
+    // bytes4(keccak256('SignedCallDataSignature(bytes)'))
+    bytes4 constant private SIGNATURE_SELECTOR = 0xf86d1d92;
+
+    /// @dev Try to parse potentially signed calldata into its hash and signature
+    ///      components. Signed calldata has signature data appended to it.
+    /// @param callData the raw call data.
+    /// @return callDataHash The hash of the signed callData, or 0x0 if no signature
+    ///         is present.
+    /// @return signature The signature bytes, if present.
+    function parseCallData(bytes memory callData)
+        internal
+        pure
+        returns (bytes32 callDataHash, bytes memory signature)
+    {
+        // Signed calldata has a 70 byte signature appended as:
+        // ```
+        //   abi.encodePacked(
+        //     callData,
+        //     bytes4(keccak256('SignedCallDataSignature(bytes)')),
+        //     signature // 66 bytes
+        //   );
+        // ```
+
+        if (
+            // Signed callData has to be at least 70 bytes long.
+            callData.length < 70 ||
+            // The bytes4 at offset -70 should equal `SIGNATURE_SELECTOR`.
+            SIGNATURE_SELECTOR != callData.readBytes4(callData.length - 70)
```

Comment:
A bit concerned about false positives arising from using the selector as a delimiter within arbitrary calldata. One way to mitigate this would be prepending the signature. Then we'd be able to guarantee no false positive, so long as we do not have a function with the same selector as a signature.

Reply:
idk I think the odds of 4 specific bytes at a specific position in a typically large payload seems pretty low? Prepending seems like it would potentially be more of a headache for integrations and our own maintenance since the calldata will be unusable unless you are aware of the signing scheme (which could change). We've also been using this scheme for metadata tagging thus far with no issues. :man_shrugging:

Reply:
I think for the current use case it's probably not a huge deal. Although I imagine this will be pointed out in the audits, so probably good to document the reasoning and worst case here in the code. I believe the worst case is just that we'll think the calldata is signed when it isn't, which would lead to a revert down the line.

Reply:
In meta-transactions there's a possibility that we could unintentionally  truncate the calldata if the 4 byte selector is found at `-70` (but it is not actually signed calldata), which might alter the results of the `abi.decode()`, and thus the parameters passed to `_transformERC20()`. This could work out detrimentally for the taker. But I think it's really unlikely and in most cases, with conventional encoding algorithms, would just cause `abi.decode()` to revert.

Reply:
Oh wait, nvm. We `abi.decode()` the raw calldata in MetaTransactions, so we should be good :zany_face:
The guy who wrote this is a genius.